libp2p/connmanager.nim

```diff
@@ -53,28 +58,42 @@ type
     muxer: Muxer
     handle: Future[void]
 
+  ConnHolder = object
+    conn: Connection
+    evtHandle: Future[void]
+
   ConnManager* = ref object of RootObj
     maxConns: int
-    # NOTE: don't change to PeerInfo here
-    # the reference semantics on the PeerInfo
-    # object itself make it susceptible to
-    # copies and mangling by unrelated code.
-    conns: Table[PeerID, HashSet[Connection]]
+    maxPeerConns: int
+    connSemaphore*: AsyncSemaphore
     muxed: Table[Connection, MuxerHolder]
+    conns: seq[ConnHolder]
     connEvents: Table[ConnEventKind, OrderedSet[ConnEventHandler]]
     peerEvents: Table[PeerEvent, OrderedSet[PeerEventHandler]]
 
 proc newTooManyConnections(): ref TooManyConnections {.inline.} =
   result = newException(TooManyConnections, "too many connections for peer")
 
 proc init*(C: type ConnManager,
-           maxConnsPerPeer: int = MaxConnectionsPerPeer): ConnManager =
-  C(maxConns: maxConnsPerPeer,
-    conns: initTable[PeerID, HashSet[Connection]](),
-    muxed: initTable[Connection, MuxerHolder]())
+           maxConns: int = MaxConnections,
+           maxPeerConns: int = MaxConnectionsPerPeer): ConnManager =
+  C(maxConns: maxConns,
+    maxPeerConns: maxPeerConns,
+    muxed: initTable[Connection, MuxerHolder](),
+    connSemaphore: AsyncSemaphore.init(maxConns))
 
 proc connCount*(c: ConnManager, peerId: PeerID): int =
-  c.conns.getOrDefault(peerId).len
+  c.conns
```

Comment:
this becomes a second source of truth, in addition to the count in the semaphore - the code is prone to have mismatches between the two because they're not updated according to the same rules

Reply:
The  semaphore counts both identified and unidentified connections whereas `countCount` returns the number of identified connections only, ie connections for which we know the peer id. This prevents inflight (unsecured/unidentified/non-upgraded) connections from exploding. The reason to do this is to make `ConnManager` the single source of truth and prevent sharing the semaphore across the switch/transport  and `ConnManager`.